chained/selector.go

```
@@ -0,0 +1,214 @@
+	// connections created via this dialer.
+	DataRecv() uint64
+
+	// Stop stops background processing for this Dialer.
+	Stop()
+
+	WriteStats(w io.Writer)
+}
+
+type Selector struct {
+	dialers []Dialer
+	// active is the index of the currently active dialer
+	active atomic.Int64
+}
+
+func NewSelector(dialers []Dialer) *Selector {
+	return &Selector{dialers: dialers}
+}
+
+func (s *Selector) Dial(network, addr string) (net.Conn, error) {
+	return s.DialContext(context.Background(), network, addr)
+}
+
+// DialContext dials the given network and address. It will cycle through the available dialers
+// until one is successful, or all have been attempted.
+func (s *Selector) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	if len(s.dialers) == 0 {
+		return nil, log.Error("no dialers available")
+	}
+
+	for i, dlr := range s.iter() {
+		if !dlr.SupportsAddr(network, addr) {
+			log.Debugf("dialer %s does not support %s %s", dlr.Label(), network, addr)
+			continue
+		}
+
+		log.Debugf("dialing %s %s with dialer %s", network, addr, dlr.Label())
+		conn, err := s.dialWithDialer(ctx, dlr, network, addr)
+		if err == nil {
+			log.Debugf("dialer %s successfully dialed %s %s", dlr.Label(), network, addr)
+			s.active.Store(int64(i))
+			return conn, nil
+		}
+
+		select {
+		case <-ctx.Done():
+			err := fmt.Errorf("dialing %s %s: %v", network, addr, ctx.Err())
+			log.Error(err)
+			return nil, err
+		default:
+		}
+
+		log.Errorf("dialer %s failed to dial %s %s: %v", dlr.Label(), network, addr, err)
+	}
+
+	return nil, fmt.Errorf("failed to dial %s %s", network, addr)
+}
+
+func (s *Selector) dialWithDialer(ctx context.Context, dlr Dialer, network, addr string) (net.Conn, error) {
+	for {
+		conn, failedUpstream, err := dlr.DialContext(ctx, network, addr)
+		if err == nil {
+			return conn, nil
+		}
+
+		select {
+		case <-ctx.Done():
+			log.Debugf("dialWithDialer: %v", ctx.Err())
+			return nil, err
+		default:
+		}
```

If we get rid of the loop here, this should no longer be necessary.  `dlr.DialContext` should return an error if the context dies.